### PR TITLE
Dedup responses

### LINKED ISSUE #149: Duplicate A records returned
I'm continuing to play with multi-port kubernetes services in SkyDNS, and it ALMOST works as I expect.

For this example, I have a service called `multiport.th1` which I want an A record for (1.1.1.1) and which hosts 2 ports (http and https).

```
# etcdctl set /skydns/local/kubernetes/th1/multiport/http '{"host":"1.1.1.1", "port":80}'
{"host":"1.1.1.1", "port":80}

# etcdctl set /skydns/local/kubernetes/th1/multiport/https '{"host":"1.1.1.1", "port":443}'
{"host":"1.1.1.1", "port":443}

# dig @10.0.0.10 +noall +answer +additional multiport.th1.kubernetes.local A
multiport.th1.kubernetes.local.	3600 IN	A	1.1.1.1
multiport.th1.kubernetes.local.	3600 IN	A	1.1.1.1

# dig @10.0.0.10 +noall +answer +additional multiport.th1.kubernetes.local SRV
multiport.th1.kubernetes.local.	3600 IN	SRV	10 50 443 https.multiport.th1.kubernetes.local.
multiport.th1.kubernetes.local.	3600 IN	SRV	10 50 80 http.multiport.th1.kubernetes.local.
https.multiport.th1.kubernetes.local. 3600 IN A	1.1.1.1
http.multiport.th1.kubernetes.local. 3600 IN A	1.1.1.1
```

The troubling issue here is that 2 A records are returned.

### PR DESCRIPTION
**Pull request by miekg**
Multiple idential RR records should be dedupped so that only one is
left.
I.e.

multiport.th1.kubernetes.local. 3600 IN A   1.1.1.1
multiport.th1.kubernetes.local. 3600 IN A   1.1.1.1

Should return:

multiport.th1.kubernetes.local. 3600 IN A   1.1.1.1

Note: I believe the clients don't really care about this and stuff just
works, but a good server should not be sending out invalid messages.

Also added test case that checks for this. We also dedup the additional
section, but for now there is no test for this.

Fixes #149

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -711,3 +715,53 @@ func (s *server) RoundRobin(rrs []dns.RR) {
...
+	}
+	// Only is our map is smaller than the #RR in the answer section we should reset the RRs
+	// in the section it self
+	if len(ma) < len(m.Answer) {
+		i := 0
+		for _, v := range ma {
+			m.Answer[i] = v
+			i++
+		}
+		m.Answer = m.Answer[:len(ma)]
```
Comment (thockin) [inline server/server.go]: Does this handle the case of multiple multiples?  Seems like it is worth a testcase

```
# dig @10.0.0.10 +noall +answer +additional +ignore svc.ns.kubernetes.local A
svc.ns.kubernetes.local. 3600	IN	A	3.3.3.3
svc.ns.kubernetes.local. 3600	IN	A	2.2.2.2
svc.ns.kubernetes.local. 3600	IN	A	3.3.3.3
svc.ns.kubernetes.local. 3600	IN	A	2.2.2.2
svc.ns.kubernetes.local. 3600	IN	A	1.1.1.1
svc.ns.kubernetes.local. 3600	IN	A	1.1.1.1
```

Comment (miekg) [inline server/server.go]: [ Quoting <notifications@github.com> in "Re: [skydns] Dedup responses (#153)..." ]
>> +	// in the section it self
>> +	if len(ma) < len(m.Answer) {
>> +		i := 0
>> +		for _, v := range ma {
>> +			m.Answer[i] = v
>> +			i++
>> +		}
>> +		m.Answer = m.Answer[:len(ma)]
>
>Does this handle the case of multiple multiples?  Seems like it is worth a testcase

Don't see why not...

>```
># dig @10.0.0.10 +noall +answer +additional +ignore svc.ns.kubernetes.local A
>svc.ns.kubernetes.local. 3600	IN	A	3.3.3.3
>svc.ns.kubernetes.local. 3600	IN	A	2.2.2.2
>svc.ns.kubernetes.local. 3600	IN	A	3.3.3.3
>svc.ns.kubernetes.local. 3600	IN	A	2.2.2.2
>svc.ns.kubernetes.local. 3600	IN	A	1.1.1.1
>svc.ns.kubernetes.local. 3600	IN	A	1.1.1.1
>```

Test == good. Added one.

/Miek

--
Miek Gieben

**Comment by thockin** [comment]
LGTM - happy to re-verify once you rebase and commit :)

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [skydns] Dedup responses (#153)..." ]
>LGTM - happy to re-verify once you rebase and commit :)

done

/Miek

--
Miek Gieben

**Comment by miekg** [comment]
merged.

**Comment by thockin** [comment]
Tested and confirmed!  Thanks for the fix.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [skydns] Dedup responses (#153)..." ]
>Tested and confirmed!  Thanks for the fix.

cool.

now please close some issues :-)

**Pull request closed without merging**